### PR TITLE
Fix infinite accessibility loop when using ToolStripControlHost 

### LINKED ISSUE #4835: ToolStripControlHost with a custom control freezes UIA clients
* .NET Core Version: 6.0

* Have you experienced this same bug with .NET Framework?: Yes

* Original bug: [1313197](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1313197?src=WorkItemMention&src-action=artifact_link)

**Problem description:**
- If a ToolStrip has ToolStripControlHost item, Inspect can't show its accessbiility tree and freezes:
hMfSttv4cb

**Expected behavior:**
- Inspect shows a correct accessbiility tree for a ToolStrip with a ToolStripControlHost:
image

**Minimal repro:**
- Create a .NET 6.0 app
- Add a ToolStrip to the app
- Create ToolStripControlHost with some control (eg. `new ToolStripControlHost(new RadioButton())`) and add it to the ToolStrip items
- Add several any items to ToolStrip (optional)
- Run the app and check the accessibility tree using Inspect

### PR DESCRIPTION
**Pull request by vladimir-krestov**
Fixes #4835
Original bug: [1313197](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1313197?src=WorkItemMention&src-action=artifact_link)

## Proposed changes

- Add a test app for ToolStrips 
- Implement FragmentRoot and FragmentNavigate for all controls to return correct value if a control is part of a ToolStrip
- Remove duplicate accessibility instances of native controls (that don't support UIA) from an accessibility tree 

## Customer Impact

- Accessibility tools don't freeze when navigate through a ToolStrip with a ToolStripControlHost 

## Regression? 

- Yes (from .NET Framework 4.7.2)

## Risk

- Minimal

### Before

- An accessibility tool (eg. Inspect) freezes on a ToolStrip if it has a ToolStripControlHost:
hMfSttv4cb

### After

- An accessibility tool (eg. Inspect) works correctly and see the correct accessibility tree:
image

## Test methodology 

- CTI
- Manual
- Unit tests

## Accessibility testing  

- Using Inspect and Narrator

## Test environment(s) 

- dotnet SDK Version: 6.0.100-preview.1.21101.5
- Microsoft Windows [Version 10.0.19042.928]

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/4850)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs [hunk 1]
```diff
@@ -206,6 +216,38 @@ internal int GetChildOverflowFragmentCount()
                 return _owningToolStrip.OverflowItems.Count;
             }
 
+            private AccessibleObject GetNextSupportedItem(int index, ToolStripItemCollection items, UiaCore.NavigateDirection direction)
+            {
+                switch (direction)
+                {
+                    case UiaCore.NavigateDirection.FirstChild:
+                    case UiaCore.NavigateDirection.NextSibling:
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs:224]: Are toolstrip items considered toolstrip's siblings? Mayme this case should go to the base class?

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs [hunk 2]
```diff
@@ -157,53 +157,59 @@ public override int GetChildCount()
...
+                    if (item.Available && item.Alignment == ToolStripItemAlignment.Right)
+                    {
+                        return GetItemAccessibleObject(item);
                     }
                 }
 
-                for (int index = 0; index < childFragmentCount; index++)
+                return null;
+
+                AccessibleObject GetItemAccessibleObject(ToolStripItem item)
```
Comment (vladimir-krestov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs:199]: Make it static

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs [hunk 3]
```diff
@@ -37,6 +37,19 @@ internal ControlAccessibleObject(Control ownerControl, int accObjId)
...
+            ///  its accessible object should provide info about the owning ToolStrip and items-siblings
+            ///  to build a correct ToolStrip accessibility tree.
+            /// </remark>
+            internal override UiaCore.IRawElementProviderFragment? FragmentNavigate(UiaCore.NavigateDirection direction)
+                => Owner.ToolStripControlHost is ToolStripControlHost host
+                    && (direction == UiaCore.NavigateDirection.Parent
+                       || direction == UiaCore.NavigateDirection.PreviousSibling
+                       || direction == UiaCore.NavigateDirection.NextSibling)
+                           ? host.AccessibilityObject.FragmentNavigate(direction)
+                           : base.FragmentNavigate(direction);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs:51]: I'm not sure if it is worth trying to inline this method, it's getting hard to read.



            internal override UiaCore.IRawElementProviderFragment? FragmentNavigate(UiaCore.NavigateDirection direction)                =&gt; Owner.ToolStripControlHost is ToolStripControlHost host                    &amp;&amp; (direction == UiaCore.NavigateDirection.Parent                       || direction == UiaCore.NavigateDirection.PreviousSibling                       || direction == UiaCore.NavigateDirection.NextSibling)                           ? host.AccessibilityObject.FragmentNavigate(direction)                           : base.FragmentNavigate(direction);



              =&gt; Owner.ToolStripControlHost is ToolStripControlHost host                    &amp;&amp; (direction == UiaCore.NavigateDirection.Parent                       || direction == UiaCore.NavigateDirection.PreviousSibling                       || direction == UiaCore.NavigateDirection.NextSibling)                           ? host.AccessibilityObject.FragmentNavigate(direction)                           : base.FragmentNavigate(direction);



              =&gt; Owner.ToolStripControlHost is ToolStripControlHost host                    &amp;&amp; (direction == UiaCore.NavigateDirection.Parent                       || direction == UiaCore.NavigateDirection.PreviousSibling                       || direction == UiaCore.NavigateDirection.NextSibling)                           ? host.AccessibilityObject.FragmentNavigate(direction)                           : base.FragmentNavigate(direction);

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs [hunk 4]
```diff
@@ -39,6 +41,103 @@ public void ControlAccessibleObject_Ctor_ControlWithoutHandle()
             Assert.Equal(string.Empty, accessibleObject.Value);
         }
 
+        public static IEnumerable<object[]> TODO()
+        {
+            //yield return new object[] { new MarshalingControl() }; // Creates Handle
+            yield return new object[] { new ParkingWindow() };
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:47]: Will you be able to dispose of these controls?

Comment (RussKie) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:47]: We should avoid using disposable object in theories, https://github.com/dotnet/winforms/blob/main/docs/testing.md#theory-tests

Comment (vladimir-krestov) [reply src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs]: I reworked this, now I'm using `Type` in theories to create and dispose controls inside a test, using `ReflectionHelper`

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs (hunk 1 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs:240]: Added comments to this method to clarify that

**Comment by RussKie** [comment]
```
Process terminated. Assertion Failed
ComboBox's handle hasn't been created
   at System.NoAssertContext.NoAssertListener.Fail(String message, String detailMessage) in /_/src/System.Windows.Forms.Primitives/tests/TestUtilities/NoAssertContext.cs:line 105
   at System.Diagnostics.TraceInternal.Fail(String message, String detailMessage) in System.Diagnostics.TraceSource.dll:token 0x60000b5+0x30
   at System.Diagnostics.TraceInternal.TraceProvider.Fail(String message, String detailMessage) in System.Diagnostics.TraceSource.dll:token 0x60000ca+0x0
   at System.Diagnostics.Debug.Fail(String message, String detailMessage) in System.Private.CoreLib.dll:token 0x6005eff+0x0
   at System.Windows.Forms.ComboBox.ChildAccessibleObject..ctor(ComboBox owner, IntPtr handle) in /_/src/System.Windows.Forms/src/System/Windows/Forms/ComboBox.ChildAccessibleObject.cs:line 19
   at System.Windows.Forms.ToolStripComboBox.ToolStripComboBoxControl.ToolStripComboBoxControlAccessibleObject..ctor(ToolStripComboBoxControl toolStripComboBoxControl) in /_/src/System.Windows.Forms/src/System/Windows/Forms/ToolStripComboBox.ToolStripComboBoxControl.ToolStripComboBoxControlAccessibleObject.cs:line 22
   at System.Windows.Forms.ToolStripComboBox.ToolStripComboBoxControl.CreateAccessibilityInstance() in /_/src/System.Windows.Forms/src/System/Windows/Forms/ToolStripComboBox.ToolStripComboBoxControl.cs:line 45
   at System.Windows.Forms.Control.get_AccessibilityObject() in /_/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs:line 504
   at System.Windows.Forms.Tests.ToolStripComboBox_ToolStripComboBoxControl_ToolStripComboBoxControlAccessibleObjectTests.ToolStripComboBoxControlAccessibleObject_FragmentNavigate_ChildrenAreExpected() in /_/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ToolStripComboBox.ToolStripComboBoxControl.ToolStripComboBoxControlAccessibleObjectTests.cs:line 80
```

File: src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs [hunk 5]
```diff
@@ -541,6 +563,14 @@ internal override bool RaiseAutomationPropertyChangedEvent(UiaCore.UIA propertyI
                 }
             }
 
+            /// <remark>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Control.ControlAccessibleObject.cs:566]: We use remarks only for xml-docs. In this case normal `//` comments are sufficient.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs [hunk 6]
```diff
@@ -1313,6 +1313,221 @@ public void ControlAccessibleObject_GetPropertyValue_ControlTypeProperty_Returns
...
+        }
+
+        [WinFormsTheory]
+        [MemberData(nameof(ControlAccessibleObject_TestData))]
+        public void ControlAccessibleObject_FragmentRoot_IsToolStrip_IfControlIsUsedAsToolStripItem(Type type)
+        {
+            using NoAssertContext context = new();
+            using ToolStrip toolStrip = new();
+            using Control control = ReflectionHelper.InvokePublicConstructor<Control>(type);
+            ToolStripControlHost host = new(control);
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:1340]: Dispose?
```suggestion
            using ToolStripControlHost host = new(control);
```

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs [hunk 7]
```diff
@@ -1313,6 +1313,221 @@ public void ControlAccessibleObject_GetPropertyValue_ControlTypeProperty_Returns
...
+            AccessibleObject expected = label.AccessibilityObject;
+
+            Assert.Equal(expected, actual);
+            Assert.True(control.IsHandleCreated);
+            Assert.True(toolStrip.IsHandleCreated);
+
+            control.Dispose();
+        }
+
+        /// <remark>
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:1519]: ditto comment

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs [hunk 8]
```diff
@@ -1313,6 +1313,221 @@ public void ControlAccessibleObject_GetPropertyValue_ControlTypeProperty_Returns
...
+        {
+            using NoAssertContext context = new();
+            using ToolStrip toolStrip = new();
+            using Control control = ReflectionHelper.InvokePublicConstructor<Control>(type);
+            ToolStripControlHost host = new(control);
+
+            toolStrip.CreateControl();
+            control.CreateControl();
+
+            if (!CanBeAddedToToolStrip(control))
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:1345]: Suggest adding a comment here (and other callsites) to the effect of: "A number of control can't be added to ToolStrip. Double verify it is still the case, and exit."

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs [hunk 9]
```diff
@@ -1313,6 +1313,221 @@ public void ControlAccessibleObject_GetPropertyValue_ControlTypeProperty_Returns
...
+            {
+                Assert.Throws<ArgumentException>(() => toolStrip.Items.Add(host));
+                control.Dispose();
```
Comment (RussKie) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/Control.ControlAccessibleObjectTests.cs:1348]: I don't think this is necessary, since you declared as `using Control control = ...`

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs [hunk 10]
```diff
@@ -338,6 +412,11 @@ public override AccessibleRole Role
                 }
             }
 
+            private bool ShouldItemBeSkipped(Control hostedControl)
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs:415]: What do you think of this name?
```suggestion
            private bool IsValidFragment(Control hostedControl)
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs [hunk 11]
```diff
@@ -206,6 +222,64 @@ internal int GetChildOverflowFragmentCount()
                 return _owningToolStrip.OverflowItems.Count;
             }
 
+            /// <summary>
+            ///  Returns the next or the previous ToolStrip item, that is not a ToolStripControlHost
+            ///  or is a control, that supports UIA providers. In fact, using this method removes hosted ToolStrip items,
+            ///  that are native controls (doesn't support UIA providers, their accessible objects is Windows responsability),
+            ///  from the accessibility tree. It's necessary, because hosted native controls internally add accessible objects
+            ///  to the accessibility tree. To avoid duplicates, remove hosted items with native accessibility objects from the tree.
+            /// </summary>
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs:231]: Few issues with this comment.

1. "_... that is not a ToolStripControlHost or is a control, that supports UIA providers..._" - statements like "this is not A or is B" are very hard to parse.
2. Spelling errors

```suggestion
            /// <summary>
            ///  Returns the next or the previous ToolStrip item, that is considered a valid navigation fragment (e.g. a control,
            ///  that supports UIA providers and not a <see cref="ToolStripControlHost"/>).
            ///  This method removes hosted ToolStrip items that are native controls (their accessible objects are provided by Windows)
            ///  from the accessibility tree. It's necessary, because hosted native controls add accessible objects
            ///  to the accessibility tree, and thus create duplicated. To avoid duplicates, remove hosted items with native accessibility objects from the tree.
            /// </summary>
```

Please reformat as appropriate.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs (hunk 10 above)
Comment (vladimir-krestov) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.ToolStripAccessibleObject.cs:415]: Hm, looks not good to me. I wouldn't say that some item are not valid. They just should be skipped to avoid duplications. But the are valid. I would keep it as is. Maybe there is possible another better name.

**Review by RussKie — APPROVED** [review]
